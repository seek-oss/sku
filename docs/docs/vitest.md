# Vitest

[Vitest] is a testing framework that supports ESM out-of-the-box, integrates with the Vite ecosystem and has a similar API to Jest.
These features make it a great replacement for Jest in `sku` applications, especially given [Jest's current limitations with ESM].
Sku supports Vitest as an alternate to Jest since v15

**We recommend using vitest with the vite bundler and in an ESM project**, however these are not requirements.

## Migrating to Vitest

Migrating to Vitest should be fairly straightforward.
Before making any changes, it is highly recommend to read [the Vitest documentation](https://vitest.dev/guide/) to familiarise yourself with the API and features, as well as to better understand the differences between Jest and Vitest.

To enable Vitest in `sku`, first install the required dependencies:

```bash
pnpm add -D vitest @sku-lib/vitest
```

Then, configure [`testRunner`][test runner] in your `sku` config:

```typescript
// sku.config.ts
import type { SkuConfig } from 'sku';

export default {
  testRunner: 'vitest',
  ...
} satisfies SkuConfig;
```

`sku` will now invoke [the `vitest` CLI][Vitest CLI] instead of the `jest` CLI when running `sku test`.

#### Key differences between Vitest and Jest

Vitest has strong compatibility with the Jest API, however it still has differences that may affect your tests.
Differences that affect sku projects are listed here.
The full list is documented in the Vitest documentation: [Migrating from Jest to Vitest].

To automate most of the migration process, a codemod is available.
Note that additional changes may still be required after running this codemod.

```sh
pnpm dlx @sku-lib/codemod jest-to-vitest .
```

**Default Watch Mode**

`vitest` defaults to watch mode when running tests.
To run tests without watch mode you can use the `--run` flag:

```json
// package.json
{
  "scripts": {
    // ...
    "test": "sku test --run"
  }
}
```

Watch mode won't trigger in CI environments, so it's safe to omit the flag in your pipeline.

**Code Coverage Dependencies**

Vitest does not install code coverage dependencies by default.
To collect code coverage, pass the `--coverage` flag to the `sku test` command and install the `@vitest/coverage-v8` package:

```bash
pnpm add -D @vitest/coverage-v8
```

**Testing Library Matchers**

If your test setup file includes an import for `@testing-library/jest-dom`, you may need to change this to `@testing-library/jest-dom/vitest`:

```diff
// test-setup.ts
- import '@testing-library/jest-dom';
+ import '@testing-library/jest-dom/vitest';
```

**Globals Disabled**

Jest enables global APIs such as `it`, `describe`, `beforeAll`, etc., by default.
Vitest does not.
Sku adopts the Vitest defaults, meaning you will need to explicitly import these test functions.
This change results in a cleaner global namespace, better type safety, clearer dependencies and is more consistency with modern javascript practices.

```diff
+ import { describe, expect, it } from 'vitest';
```

Be aware that since globals are disabled, some common libraries like `testing-library` will not run auto DOM cleanup.
If using these libraries, you will need to add cleanup to your configured `setupTests` file.

```diff
import '@testing-library/jest-dom/vitest';

+ import { cleanup } from '@testing-library/react';
+ import { afterEach } from 'vitest';

+ afterEach(cleanup)
```

[Vitest]: https://vitest.dev/
[Jest's current limitations with ESM]: https://jestjs.io/docs/ecmascript-modules
[test runner]: ./docs/configuration.md#testrunner
[Vitest CLI]: https://vitest.dev/guide/cli.html
[codemod]: https://codemod.com/registry/jest-vitest
[Migrating from Jest to Vitest]: https://vitest.dev/guide/migration.html#jest
[Vitest globals]: https://vitest.dev/config/#globals
