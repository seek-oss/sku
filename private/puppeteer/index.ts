export { getAppSnapshot } from './appSnapshot.ts';
export { appSnapshotSerializer } from './appSnapshotSerializer.ts';
export { htmlSnapshotSerializer } from './htmlSnapshotSerializer.ts';
export { cssSnapshotSerializer } from './cssSnapshotSerializer.ts';
