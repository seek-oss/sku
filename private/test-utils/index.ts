export { ListExternalsWebpackPlugin } from './ListExternalsWebpackPlugin.ts';
export { dirContentsToObject } from './dirContentsToObject.ts';
export { makeStableHashes, makeStableViteHashes } from './skuConfig.ts';
export { getStoryPage, getTextContentFromFrameOrPage } from './storybook.ts';
export { getPort } from './getPort.ts';
