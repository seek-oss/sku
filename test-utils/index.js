// @ts-check
export { ListExternalsWebpackPlugin } from './ListExternalsWebpackPlugin.js';
export { getAppSnapshot } from './appSnapshot.js';
export { startAssetServer } from './assetServer.js';
export { dirContentsToObject } from './dirContentsToObject.js';
export { run, runSkuScriptInDir } from './process.ts';
export { makeStableHashes } from './skuConfig.js';
export { getStoryPage, getTextContentFromFrameOrPage } from './storybook.js';
export { waitForUrls } from './waitForUrls.js';
