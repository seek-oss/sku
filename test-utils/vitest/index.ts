export { getAppSnapshot } from './appSnapshot.ts';
