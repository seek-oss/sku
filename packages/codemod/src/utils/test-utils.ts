import { readFile } from 'node:fs/promises';
import path from 'node:path';
import { expect } from 'vitest';
import { isValidTransform, type Transform } from './types.js';

type FileInfo = {
  source: string;
  path: string;
};

type InlineTestProps = {
  transform: Transform;
  input: FileInfo;
  expectedOutput: string;
};

type TestProps = {
  dirName: string;
  transformName: string;
  fixtureName?: string;
  testOptions?: { extension?: string };
};

const runInlineTest = async ({
  transform,
  input,
  expectedOutput,
}: InlineTestProps) => {
  if (!isValidTransform(transform)) {
    throw new Error('Invalid transform function provided');
  }

  const output = transform(input.source);
  if (!output) {
    throw new Error('Transform failed');
  }

  expect(output.trim()).toEqual(expectedOutput.trim());

  return output;
};

export const runTest = async ({
  dirName,
  transformName,
  fixtureName,
  testOptions = {},
}: TestProps) => {
  const testFixtureName = fixtureName || transformName;

  const fixtureDir = path.join(dirName, '..', '__testfixtures__');
  const inputPath = path.join(
    fixtureDir,
    `${testFixtureName}.input.${testOptions.extension}`,
  );

  // Assumes transform is one level up from __tests__ directory
  const transformImport = import(path.join(dirName, '..', transformName));
  const sourceRead = readFile(inputPath, 'utf8');
  const expectedOutputRead = readFile(
    path.join(fixtureDir, `${testFixtureName}.output.${testOptions.extension}`),
    'utf8',
  );

  const [{ transform }, source, expectedOutput] = await Promise.all([
    transformImport,
    sourceRead,
    expectedOutputRead,
  ]);

  return runInlineTest({
    transform,
    input: {
      path: inputPath,
      source,
    },
    expectedOutput,
  });
};

export const runNoChangeTest = async ({
  dirName,
  transformName,
  fixtureName,
  testOptions = {},
}: TestProps) => {
  const inputPath = path.join(
    dirName,
    '..',
    '__testfixtures__',
    `${fixtureName}.unchanged.${testOptions.extension}`,
  );

  const sourceRead = readFile(inputPath, 'utf8');
  const transformImport = import(path.join(dirName, '..', transformName));

  const [{ transform }, source] = await Promise.all([
    transformImport,
    sourceRead,
  ]);
  const output = await transform(source);

  expect(output).toEqual(null);
};
