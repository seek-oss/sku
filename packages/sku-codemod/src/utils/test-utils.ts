import { readFile } from 'node:fs/promises';
import path from 'node:path';
import { expect } from 'vitest';

type FileInfo = {
  source: string;
  path: string;
};

const runInlineTest = async (
  transform: (source: string) => string | false,
  input: FileInfo,
  expectedOutput: string,
) => {
  const output = await transform(input.source);
  if (!output) {
    throw new Error('Transform failed');
  }
  const expectation = (o: string) =>
    expect(o.trim()).toEqual(expectedOutput.trim());
  expectation(output);
  return output;
};

export const runTest = async (
  dirName: string,
  transformName: string,
  testFilePrefix?: string,
  testOptions: Record<string, unknown> = {},
) => {
  let localTestFilePrefix = testFilePrefix;
  if (!testFilePrefix) {
    localTestFilePrefix = transformName;
  }

  // Assumes transform is one level up from __tests__ directory
  const transform = (await import(path.join(dirName, '..', transformName)))
    .transform;
  const fixtureDir = path.join(dirName, '..', '__testfixtures__');
  const inputPath = path.join(
    fixtureDir,
    `${localTestFilePrefix}.input.${testOptions.extension}`,
  );
  const source = await readFile(inputPath, 'utf8');
  const expectedOutput = await readFile(
    path.join(
      fixtureDir,
      `${localTestFilePrefix}.output.${testOptions.extension}`,
    ),
    'utf8',
  );
  return runInlineTest(
    transform,
    {
      path: inputPath,
      source,
    },
    expectedOutput,
  );
};

export const runNoChangeTest = async (
  dirName: string,
  transformName: string,
  testFilePrefix: string,
  testOptions: Record<string, unknown>,
) => {
  const transform = (await import(path.join(dirName, '..', transformName)))
    .transform;
  const inputPath = path.join(
    dirName,
    '..',
    '__testfixtures__',
    `${testFilePrefix}.unchanged.${testOptions.extension}`,
  );
  const source = await readFile(inputPath, 'utf8');
  const output = await transform(source);
  expect(output).toEqual(false);
};
