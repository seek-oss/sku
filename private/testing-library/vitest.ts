// Re-export vitest setup from cli-testing-library
import 'cli-testing-library/vitest';
