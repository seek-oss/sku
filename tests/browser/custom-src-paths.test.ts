import { describe, beforeAll, it, expect, afterAll } from 'vitest';
import { dirContentsToObject, getPort } from '@sku-private/test-utils';

import { getAppSnapshot } from '@sku-private/puppeteer';

import {
  bundlers,
  type BundlerValues,
  cleanup,
  skipCleanup,
  scopeToFixture,
} from '@sku-private/testing-library';

const { sku, fixturePath } = scopeToFixture('custom-src-paths');

describe('custom-src-paths', () => {
  describe.for(bundlers)('bundler %s', (bundler) => {
    describe('start', async () => {
      it('should start a development server', async () => {
        const port = await getPort();
        const url = `http://localhost:${port}`;

        const args: BundlerValues<string[]> = {
          vite: [
            '--config',
            'sku.config.vite.ts',
            '--strict-port',
            `--port=${port}`,
          ],
          webpack: ['--strict-port', `--port=${port}`],
        };

        const start = await sku('start', args[bundler]);
        expect(
          await start.findByText('Starting development server'),
        ).toBeInTheConsole();

        const snapshot = await getAppSnapshot({ url, expect });
        expect(snapshot).toMatchSnapshot();
      });
    });

    describe('build', async () => {
      const port = await getPort();
      const url = `http://localhost:${port}`;
      const portArgs = ['--strict-port', `--port=${port}`];

      const args: BundlerValues<string[]> = {
        vite: ['--config', 'sku.config.vite.ts'],
        webpack: [],
      };

      beforeAll(async () => {
        const build = await sku('build', args[bundler]);
        expect(await build.findByText('Build complete')).toBeInTheConsole();

        const serve = await sku('serve', portArgs);
        expect(await serve.findByText('Server started')).toBeInTheConsole();
      });

      afterAll(cleanup);

      it('should generate the expected files', async ({ task }) => {
        skipCleanup(task.id);
        const files = await dirContentsToObject(fixturePath('dist'));
        expect(files).toMatchSnapshot();
      });

      it('should create valid app', async ({ task }) => {
        skipCleanup(task.id);

        const app = await getAppSnapshot({ url, expect });
        expect(app).toMatchSnapshot();
      });
    });
  });

  describe('format', () => {
    it('should format successfully', async () => {
      const format = await sku('format');
      expect(await format.findByText('Formatting complete')).toBeInTheConsole();
    });
  });

  describe('lint', () => {
    it('should lint successfully', async () => {
      const lint = await sku('lint');
      expect(await lint.findByText('Linting complete')).toBeInTheConsole();
    });
  });
});
