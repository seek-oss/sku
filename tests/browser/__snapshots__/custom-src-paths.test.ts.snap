// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`custom-src-paths > bundler vite > build > should create valid app 1`] = `
===================================================================
--- sourceHtml
+++ clientHtml
<!DOCTYPE html>
<html>
  <head>
    <meta charset="UTF-8">
    <title>
      hello-world
    </title>
    <meta
      name="viewport"
      content="width=device-width, initial-scale=1"
    >
    <link
      rel="modulepreload"
      href="/some-static-place/vite-client.js"
      crossorigin
    >
  </head>
  <body>
    <div id="app">
      <div>
        Hello custom src paths
      </div>
    </div>
    <script
      type="module"
      src="/some-static-place/vite-client.js"
    >
    </script>
  </body>
</html>
`;

exports[`custom-src-paths > bundler vite > build > should generate the expected files 1`] = `
{
  "index.html": SCRIPTS: [
    "/some-static-place/vite-client.js",
  ]
CSS: []
SOURCE HTML: <!DOCTYPE html>
<html>
  <head>
    <meta charset="UTF-8">
    <title>
      hello-world
    </title>
    <meta
      name="viewport"
      content="width=device-width, initial-scale=1"
    >
    <link
      rel="modulepreload"
      href="SCRIPTS[0]"
      crossorigin
    >
  </head>
  <body>
    <div id="app">
      <div>
        Hello custom src paths
      </div>
    </div>
    <script
      type="module"
      src="SCRIPTS[0]"
    >
    </script>
  </body>
</html>,
  "vite-client.js": "CONTENTS IGNORED IN SNAPSHOT TEST",
  "vite-client.js.map": "CONTENTS IGNORED IN SNAPSHOT TEST",
}
`;

exports[`custom-src-paths > bundler vite > start > should start a development server 1`] = `
===================================================================
--- sourceHtml
+++ clientHtml
@@ -1,60 +1,60 @@
 <!DOCTYPE html>
 <html>
   <head>
     <script type="module">
       import { injectIntoGlobalHook } from "/@react-refresh";
 injectIntoGlobalHook(window);
 window.$RefreshReg$ = () => {};
 window.$RefreshSig$ = () => (type) => type;
     </script>
     <script
       type="module"
       src="/@vite/client"
     >
     </script>
     <meta charset="UTF-8">
     <title>
       hello-world
     </title>
     <meta
       name="viewport"
       content="width=device-width, initial-scale=1"
     >
     <script
       type="module"
-      src="/@id/__x00__/index.html?html-proxy&index=0.js"
+      src="/@id/__x00__/index.html?html-proxy&amp;index=0.js"
     >
     </script>
     <script
       type="module"
-      src="/@id/__x00__/index.html?html-proxy&index=1.js"
+      src="/@id/__x00__/index.html?html-proxy&amp;index=1.js"
     >
     </script>
     <link
       rel="stylesheet"
       href="/@id/__x00__virtual:ssr-css.css"
       data-ssr-css
     >
     <script type="module">
       import { createHotContext } from "/@vite/client";
               const hot = createHotContext("/__clear_ssr_css");
               hot.on("vite:afterUpdate", () => {
                 document
                   .querySelectorAll("[data-ssr-css]")
                   .forEach(node => node.remove());
               });
     </script>
   </head>
   <body>
     <div id="app">
       <div>
         Hello custom src paths
       </div>
     </div>
     <script
       type="module"
       src="{cwd}/packages/sku/dist/services/vite/entries/vite-client.js"
     >
     </script>
   </body>
 </html>
\\ No newline at end of file
`;

exports[`custom-src-paths > bundler webpack > build > should create valid app 1`] = `
===================================================================
--- sourceHtml
+++ clientHtml
<!DOCTYPE html>
<html>
  <head>
    <meta charset="UTF-8">
    <title>
      hello-world
    </title>
    <meta
      name="viewport"
      content="width=device-width, initial-scale=1"
    >
    <link
      data-chunk="main"
      rel="preload"
      as="script"
      href="/some-static-place/runtime.js"
    >
    <link
      data-chunk="main"
      rel="preload"
      as="script"
      href="/some-static-place/2.js"
    >
    <link
      data-chunk="main"
      rel="preload"
      as="script"
      href="/some-static-place/main.js"
    >
  </head>
  <body>
    <div id="app">
      <div>
        Hello custom src paths
      </div>
    </div>
    <script
      id="__LOADABLE_REQUIRED_CHUNKS__"
      type="application/json"
    >
      []
    </script>
    <script
      id="__LOADABLE_REQUIRED_CHUNKS___ext"
      type="application/json"
    >
      {"namedChunks":[]}
    </script>
    <script
      async
      data-chunk="main"
      src="/some-static-place/runtime.js"
    >
    </script>
    <script
      async
      data-chunk="main"
      src="/some-static-place/2.js"
    >
    </script>
    <script
      async
      data-chunk="main"
      src="/some-static-place/main.js"
    >
    </script>
  </body>
</html>
`;

exports[`custom-src-paths > bundler webpack > build > should generate the expected files 1`] = `
{
  "2.js": "CONTENTS IGNORED IN SNAPSHOT TEST",
  "2.js.map": "CONTENTS IGNORED IN SNAPSHOT TEST",
  "index.html": SCRIPTS: [
    "/some-static-place/runtime.js",
    "/some-static-place/2.js",
    "/some-static-place/main.js",
  ]
CSS: []
SOURCE HTML: <!DOCTYPE html>
<html>
  <head>
    <meta charset="UTF-8">
    <title>
      hello-world
    </title>
    <meta
      name="viewport"
      content="width=device-width, initial-scale=1"
    >
    <link
      data-chunk="main"
      rel="preload"
      as="script"
      href="SCRIPTS[0]"
    >
    <link
      data-chunk="main"
      rel="preload"
      as="script"
      href="SCRIPTS[1]"
    >
    <link
      data-chunk="main"
      rel="preload"
      as="script"
      href="SCRIPTS[2]"
    >
  </head>
  <body>
    <div id="app">
      <div>
        Hello custom src paths
      </div>
    </div>
    <script
      id="__LOADABLE_REQUIRED_CHUNKS__"
      type="application/json"
    >
      []
    </script>
    <script
      id="__LOADABLE_REQUIRED_CHUNKS___ext"
      type="application/json"
    >
      {"namedChunks":[]}
    </script>
    <script
      async
      data-chunk="main"
      src="SCRIPTS[0]"
    >
    </script>
    <script
      async
      data-chunk="main"
      src="SCRIPTS[1]"
    >
    </script>
    <script
      async
      data-chunk="main"
      src="SCRIPTS[2]"
    >
    </script>
  </body>
</html>,
  "main.js": "CONTENTS IGNORED IN SNAPSHOT TEST",
  "main.js.map": "CONTENTS IGNORED IN SNAPSHOT TEST",
  "runtime.js": "CONTENTS IGNORED IN SNAPSHOT TEST",
  "runtime.js.map": "CONTENTS IGNORED IN SNAPSHOT TEST",
}
`;

exports[`custom-src-paths > bundler webpack > start > should start a development server 1`] = `
===================================================================
--- sourceHtml
+++ clientHtml
<!DOCTYPE html>
<html>
  <head>
    <meta charset="UTF-8">
    <title>
      hello-world
    </title>
    <meta
      name="viewport"
      content="width=device-width, initial-scale=1"
    >
    <link
      data-chunk="main"
      rel="preload"
      as="script"
      href="/runtime.js"
    >
    <link
      data-chunk="main"
      rel="preload"
      as="script"
      href="/2.js"
    >
    <link
      data-chunk="main"
      rel="preload"
      as="script"
      href="/main.js"
    >
  </head>
  <body>
    <div id="app">
      <div>
        Hello custom src paths
      </div>
    </div>
    <script
      id="__LOADABLE_REQUIRED_CHUNKS__"
      type="application/json"
    >
      []
    </script>
    <script
      id="__LOADABLE_REQUIRED_CHUNKS___ext"
      type="application/json"
    >
      {"namedChunks":[]}
    </script>
    <script
      async
      data-chunk="main"
      src="/runtime.js"
    >
    </script>
    <script
      async
      data-chunk="main"
      src="/2.js"
    >
    </script>
    <script
      async
      data-chunk="main"
      src="/main.js"
    >
    </script>
  </body>
</html>
`;
